### PR TITLE
Added documentation for DeletionPropagationPolicy for cleanupPolicy a…

### PR DESCRIPTION
**Pull request by ShivamJha2436**
…nd TTL-based cleanup resources

## Related issue #

This PR is a draft and needs further discussion with maintainers. No specific issue has been linked yet.

## Proposed Changes

This PR includes the following updates:

1. Expanded information on Cleanup Policies to provide more details on ClusterCleanupPolicy and CleanupPolicy types.
2. Added explanations on the deletionPropagationPolicy options for both cleanup policies and TTL-based cleanup labels, highlighting Foreground, Background, and Orphan options.

## Checklist

- [] I have read the [contributing guidelines](https://github.com/kyverno/website/blob/main/CONTRIBUTING.md).
- [] I have inspected the website preview for accuracy.
- [] I have signed off my issue.

### CODE REVIEW

File: content/en/docs/writing-policies/cleanup.md [hunk 1]
```diff
@@ -66,6 +66,50 @@ rules:
...
+      value: 2
+  schedule: "*/5 * * * *"
+  deletionPropagationPolicy: "Foreground"
+```
+
+{{% alert title="Note" color="info" %}} Since cleanup policies always operate against existing resources in a cluster, policies created with subjects, Roles, or ClusterRoles in the match/exclude block are not allowed since this information is only known at admission time. Additionally, operations[], while permitted, are ignored as the only trigger is schedule based. {{% /alert %}}
+
+Values from resources to be evaluated during a policy may be referenced with target.* similar to mutate existing rules.
+
+Because Kyverno follows the principle of least privilege, depending on the resources you wish to remove it may be necessary to grant additional permissions to the cleanup controller. Kyverno will assist in informing you if additional permissions are required by validating them at the time a new cleanup policy is installed. See the Customizing Permissions section for more details.
```
Comment (eddycharly) [inline content/en/docs/writing-policies/cleanup.md:109]: > See the Customizing Permissions section for more details.

Can this be a link to the section ?

File: content/en/docs/writing-policies/cleanup.md [hunk 2]
```diff
@@ -96,3 +140,35 @@ spec:
...
+
+### Deletion Propagation Policy in TTL-based Cleanup
+
+The deletionPropagationPolicy can also be specified for resources with a TTL-based cleanup label. This field provides flexibility in handling dependent resources when a resource reaches its expiration. For example:
+
+- **Foreground**: Ensures that all dependent resources are deleted before the resource itself is removed.
+- **Background**: Deletes the resource first, while dependents are removed asynchronously.
+- **Orphan**: Deletes the resource but leaves its dependents in place.
+
+For example, consider a Pod with the TTL label cleanup.kyverno.io/ttl: 2m. After two minutes, the Pod will be deleted, but the dependents will be handled according to the specified deletionPropagationPolicy. If the policy is set to Orphan, the Pod will be deleted, but its dependent resources will remain in the cluster.
```
Comment (eddycharly) [inline content/en/docs/writing-policies/cleanup.md:152]:
```suggestion
For example, consider a Pod with the TTL label `cleanup.kyverno.io/ttl: 2m`. After two minutes, the Pod will be deleted, but the dependents will be handled according to the specified deletionPropagationPolicy. If the policy is set to Orphan, the Pod will be deleted, but its dependent resources will remain in the cluster.
```

File: content/en/docs/writing-policies/cleanup.md (hunk 1 above)
Comment (ShivamJha2436) [reply content/en/docs/writing-policies/cleanup.md:105]: Okay sure

File: content/en/docs/writing-policies/cleanup.md [hunk 3]
```diff
@@ -66,6 +66,46 @@ rules:
   - delete
 ```
 
+### Deletion Propagation Policy
```
Comment (vishal-chdhry) [inline content/en/docs/writing-policies/cleanup.md:69]: This section should not be in the main cleanup policy body

File: content/en/docs/writing-policies/cleanup.md [hunk 4]
```diff
@@ -96,3 +136,35 @@ spec:
...
+  name: foo
+spec:
+  containers:
+  - args:
+    - sleep
+    - 1d
+    image: busybox:1.35
+    name: foo
+deletionPropagationPolicy: "Orphan"
+```
```
Comment (vishal-chdhry) [inline content/en/docs/writing-policies/cleanup.md:166]: Deletion propagation policy cannot be added to pod manifest, nor does it work with ttl label

**Review by vishal-chdhry — CHANGES_REQUESTED** [review]

File: content/en/docs/writing-policies/cleanup.md (hunk 3 above)
Comment (ShivamJha2436) [reply content/en/docs/writing-policies/cleanup.md:69]: Okay.

File: content/en/docs/writing-policies/cleanup.md [hunk 5]
```diff
@@ -96,3 +136,35 @@ spec:
...
+    cleanup.kyverno.io/ttl: 2m
+  name: foo
+spec:
+  containers:
+  - args:
+    - sleep
+    - 1d
+    image: busybox:1.35
+    name: foo
+deletionPropagationPolicy: "Orphan"
```
[the remark this answers is not in the document]
Comment (eddycharly) [reply content/en/docs/writing-policies/cleanup.md]: You didn't use the right label :(

File: content/en/docs/writing-policies/cleanup.md [hunk 6]
```diff
@@ -96,3 +96,71 @@ spec:
...
+
+- **Foreground**: Ensures dependent resources are deleted before the primary resource is removed.
+- **Background**: Deletes the primary resource first, while dependents are removed asynchronously.
+- **Orphan**: Deletes the primary resource but leaves its dependents untouched.
+
+{{% alert title="Note" color="info" %}}
+If deletionPropagationPolicy is not set, Kyverno defers to the Kubernetes API server's default behavior, which typically handles dependents based on cluster settings.
+{{% /alert %}}
+
+### Cleanup Policy Example with deletionPropagationPolicy ###
```
Comment (eddycharly) [inline content/en/docs/writing-policies/cleanup.md:114]:
```suggestion
### Cleanup Policy Example with deletionPropagationPolicy
```

File: content/en/docs/writing-policies/cleanup.md [hunk 7]
```diff
@@ -96,3 +96,71 @@ spec:
...
+      operator: LessThan
+      value: 2
+  schedule: "*/5 * * * *"
+  deletionPropagationPolicy: "Foreground"
+```
+This policy schedules the deletion of Deployments labeled canremove: "true" with fewer than two replicas every 5 minutes, ensuring dependent resources are deleted before the Deployment itself.
+
+### TTL-Based Cleanup Example with deletionPropagationPolicy ###
+
+Resources with a cleanup.kyverno.io/ttl label can also use the deletionPropagationPolicy to manage dependent resources:
```
Comment (eddycharly) [inline content/en/docs/writing-policies/cleanup.md:144]:
```suggestion
Resources with a `cleanup.kyverno.io/ttl` label can also use the deletionPropagationPolicy to manage dependent resources:
```

File: content/en/docs/writing-policies/cleanup.md [hunk 8]
```diff
@@ -96,3 +96,71 @@ spec:
...
+
+```yaml
+apiVersion: v1
+kind: CleanupPolicy
```
Comment (eddycharly) [inline content/en/docs/writing-policies/cleanup.md:148]: Hmm, it's not a cleanup policy, it's a pod.

File: content/en/docs/writing-policies/cleanup.md [hunk 9]
```diff
@@ -96,3 +96,71 @@ spec:
 Although labeled resources are watched by Kyverno, the cleanup interval (the time resolution at which any cleanup can be performed) is controlled by a flag passed to the cleanup controller called `ttlReconciliationInterval`. This value is set to `1m` by default and can be changed if a longer resolution is required.
 
 Because this is a label, there is opportunity to chain other Kyverno functionality around it. For example, it is possible to use a Kyverno mutate rule to assign this label to matching resources. A validate rule could be written prohibiting, for example, users from the `infra-ops` group from assigning the label to resources in certain Namespaces. Or, Kyverno could generate a new resource with this label as part of the resource definition.
+
+## DeletionPropagationPolicy (Common to both)
+
+The deletionPropagationPolicy field is an optional setting available in both CleanupPolicy and TTL-based cleanup configurations. It determines how Kubernetes handles the deletion of dependent resources when the primary resource is deleted.
```
Comment (eddycharly) [inline content/en/docs/writing-policies/cleanup.md:102]:
```suggestion
The `deletionPropagationPolicy` field is an optional setting available in both CleanupPolicy and TTL-based cleanup configurations. It determines how Kubernetes handles the deletion of dependent resources when the primary resource is deleted.
```

File: content/en/docs/writing-policies/cleanup.md [hunk 10]
```diff
@@ -96,3 +96,71 @@ spec:
...
+  name: foo
+spec:
+  containers:
+  - args:
+    - sleep
+    - 1d
+    image: busybox:1.35
+    name: foo
+```
+In this example:
```
Comment (eddycharly) [inline content/en/docs/writing-policies/cleanup.md:163]:
```suggestion

In this example:
```

File: content/en/docs/writing-policies/cleanup.md [hunk 11]
```diff
@@ -96,3 +96,72 @@ spec:
...
+Resources with a `cleanup.kyverno.io/ttl` label can also use the deletionPropagationPolicy to manage dependent resources:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    cleanup.kyverno.io/ttl: 2m
+  annotations:
+    deletionPropagationPolicy: "Orphan"
```
Comment (eddycharly) [inline content/en/docs/writing-policies/cleanup.md:154]: Still not the expected annotation !

File: content/en/docs/writing-policies/cleanup.md [hunk 12]
```diff
@@ -96,3 +96,72 @@ spec:
...
+Resources with a `cleanup.kyverno.io/ttl` label can also use the deletionPropagationPolicy to manage dependent resources:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    cleanup.kyverno.io/ttl: 2m
+  annotations:
+    deleteOptions: "Orphan"
```
Comment (eddycharly) [inline content/en/docs/writing-policies/cleanup.md:154]: You wrote, the code, you know the expected annotation right ? 🤔

File: content/en/docs/writing-policies/cleanup.md [hunk 13]
```diff
@@ -86,13 +86,91 @@ metadata:
...
+    any:
+      - key: "{{ target.spec.replicas }}"
+        operator: LessThan
+        value: 2
+  schedule: "*/5 * * * *"
+  # use Foreground deletion propagation policy
+  deletionPropagationPolicy: Foreground
+```
+
+This policy schedules the deletion of Deployments labeled `canremove: "true"` with fewer than two replicas every 5 minutes, ensuring dependent resources are deleted before the Deployment itself.
```
[the remark this answers is not in the document]
Comment (eddycharly) [reply content/en/docs/writing-policies/cleanup.md]: Done

**Review by vishal-chdhry — APPROVED** [review]

**Review by vishal-chdhry — APPROVED** [review]

**Pull request merged**